engineapi/proxy.go

```diff
@@ -1,168 +1,421 @@
 package engineapi
 
 import (
+	"fmt"
 	"sync"
 	"time"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/net/context"
+	"google.golang.org/grpc/connectivity"
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
 	imtypes "github.com/longhorn/longhorn-instance-manager/pkg/types"
 	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
 
 	"github.com/longhorn/longhorn-manager/datastore"
 	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/util"
 )
 
 const (
-	ProxyPollInterval = 5 * time.Second
+	MaxProxyClientConcurrentInUse = 1000
+	DefaultProxyClientPerImage    = 3
+	MaxProxyClientPerImage        = 3 * DefaultProxyClientPerImage
+	ProxyPollInterval             = 5 * time.Second
+	ProxyClientMaxRetry           = 10
+	ProxyGCMaxRetry               = 3
 )
 
-type EngineClientProxy interface {
-	EngineClient
-	BackupTargetBinaryClient
-
-	IsGRPC() bool
-	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
-	Stop(string) error
-	Ping() error
-}
-
 type EngineClientProxyHandler struct {
-	Clients map[string]EngineClientProxy
+	proxyPools map[string]engineClientProxyPool
 
 	logger logrus.FieldLogger
+	ds     *datastore.DataStore
 
-	ds *datastore.DataStore
+	stopCh chan struct{}
 
 	lock *sync.RWMutex
+
+	resyncPeriod time.Duration
+}
+
+type engineClientProxyPool struct {
+	clients map[string]EngineClientProxy
+
+	instanceManagerName string
 }
 
-func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore) *EngineClientProxyHandler {
+func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore, resyncPeriod time.Duration, stopCh chan struct{}) *EngineClientProxyHandler {
 	return &EngineClientProxyHandler{
-		Clients: make(map[string]EngineClientProxy),
-		logger:  logger,
-		ds:      ds,
-		lock:    &sync.RWMutex{},
+		proxyPools:   make(map[string]engineClientProxyPool),
+		logger:       logger,
+		ds:           ds,
+		stopCh:       stopCh,
+		lock:         &sync.RWMutex{},
+		resyncPeriod: resyncPeriod,
 	}
 }
 
-func getLoggerForProxy(logger logrus.FieldLogger, image string) *logrus.Entry {
+func getLoggerForProxyHandler(logger logrus.FieldLogger, im *longhorn.InstanceManager) *logrus.Entry {
 	return logger.WithFields(
 		logrus.Fields{
-			"instanceManagerImage": image,
+			"instanceManager": im.Name,
+			"image":           im.Spec.Image,
 		},
 	)
 }
 
+func (h *EngineClientProxyHandler) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	h.logger.Info("Starting Longhorn engine client proxy handler")
+	defer h.logger.Info("Shutting down Longhorn engine client proxy handler")
+
+	go wait.Until(h.resync, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (h *EngineClientProxyHandler) resync() {
```

Comment:
There are 4 parts in resync. Because there is no any private methods to compose this, could you add a comment for each part? This will help for the following debugging. 

1. Garbage collect or delete outdated client
2. Request to delete outdated client
3. Request to create a new client
4. Request to create an extra client when number of client connections not enough